flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -841,6 +842,12 @@ private StreamConfig createJobVertex(Integer streamNodeId, OperatorChainInfo cha
                             },
                             serializationExecutor));
         }
+        coordinatorSerializationFutures.add(
+                FutureUtils.combineAll(serializationFutures)
+                        .thenAccept(
+                                serializedCoordinators ->
+                                        serializedCoordinators.forEach(
+                                                jobVertex::addOperatorCoordinator)));
```

Comment:
hmm. This isn't quite what I had in mind TBH; you still have multiple threads modifying the `JobGraph`; it "only" works because they touch different job vertices.

Rather I wanted the `coordinatorSerializationFutures` to return a sort of `Map<JobVertex, Collection<SerializedValue>>` that we'd wait for.

Reply:
ok, I didn't think that far. :+1:

Reply:
A simpler option would be to migrate coordinatorSerializationFutures to a `Map<JobVertex, CompletableFuture<Collection<SerializedValue>>`

Reply:
hm, yeah that's what I had in mind. But if we want to do it properly, we'd have to pass the `Collection<SerialiazedValue<OperatorCoordinator>>` as a constructor parameter rather than using the `JobVertex#addOperatorCoordinator` (which we would remove then). Am I following here correctly?

Reply:
Ok, (as always :innocent: ) forget my previous statement. I misunderstood your concern initially. I pushed another proposal in 383bb9380f82a97cb8ae3d211add64ffbaa5f023